scripts/parse_file_map.py

```diff
@@ -0,0 +1,169 @@
+#!/usr/bin/env python3
+# pylint: disable=too-many-instance-attributes,too-many-locals,unused-argument,no-self-use,wrong-import-order,unused-argument
+import logging
+import os
+from io import StringIO
+from parser import GenericParser, GroupedRow
+from typing import Dict, List
+
+import click
+
+from sample_metadata.models.sequence_status import SequenceStatus
+
+logger = logging.getLogger(__file__)
+logger.addHandler(logging.StreamHandler())
+logger.setLevel(logging.INFO)
+
+
+class Columns:
+    """Column keys for file_map"""
+
+    INDIVIDUAL_ID = 'Individual ID'
+    FILENAMES = 'Filenames'
+
+
+class SampleMapParser(GenericParser):
+    """Parser for SampleMap"""
+
+    def __init__(
+        self,
+        search_locations: List[str],
+        sample_metadata_project: str,
+        default_sequence_type='wgs',
+        default_sample_type='blood',
+        confirm=False,
+        delimeter='\t',
+    ):
+        super().__init__(
+            path_prefix=search_locations[0],
+            sample_metadata_project=sample_metadata_project,
+            default_sequence_type=default_sequence_type,
+            default_sample_type=default_sample_type,
+            delimeter=delimeter,
+            confirm=confirm,
+        )
+        self.search_locations = search_locations
+        self.filename_map = {}
+        self.populate_filename_map(self.search_locations)
+
+    def populate_filename_map(self, search_locations: List[str]):
+        """
+        ParseFileMap uses search locations based on the filename,
+        so let's prepopulate that filename_map from the search_locations!
+        """
+        self.filename_map = {}
+        for directory in search_locations:
+            for file in self.list_directory(directory):
+                file_base = os.path.basename(file)
+                if file_base in self.filename_map:
+                    logger.warning(
+                        f'File "{file}" already exists in directory map: {self.filename_map[file_base]}'
+                    )
+                    continue
+                self.filename_map[file_base] = file
+
+    def file_path(self, filename: str) -> str:
+        """
+        Get complete filepath of filename:
+        - Includes gs://{bucket} if relevant
+        - Includes path_prefix decided early on
+        """
+        if filename in self.filename_map:
+            return self.filename_map[filename]
+
+        return super().file_path(filename)
+
+    def get_sample_id(self, row: Dict[str, any]):
+        """Get external sample ID from row"""
+        external_id = row[Columns.INDIVIDUAL_ID]
+        return external_id
+
+    def get_sample_meta(self, sample_id: str, row: GroupedRow):
+        """Get sample-metadata from row"""
+        return {}
+
+    def get_sequence_meta(self, sample_id: str, row: GroupedRow):
```

Comment:
I find it super helpful to specify the return type as well, helps to read code, and also can benefit from `mypy` static analysis :)

Reply:
Thanks! I've added static types to them all.